webpack/components/Search/__tests__/search.test.js

```diff
@@ -34,5 +48,77 @@ test('Autocomplete shows on input', async (done) => {
   await waitFor(() => expect(getByText(`${suggestion}`)).toBeInTheDocument());
 
   assertNockRequest(initialScope);
+  assertNockRequest(autoSearchScope);
+  assertNockRequest(autocompleteScope, done);
+});
+
+test('autosearch turned on does not show patternfly 4 search button', async (done) => {
+  const autoSearchScope = mockSetting(nockInstance, AUTOSEARCH_WHILE_TYPING, true);
+  const autocompleteScope = mockAutocomplete(nockInstance, endpoint);
+
+  const { queryByLabelText } = renderWithRedux(<Search {...props} />);
+
+  await waitFor(() => expect(queryByLabelText(searchButtonLabel)).not.toBeInTheDocument());
+
+  assertNockRequest(autocompleteScope);
+  assertNockRequest(autoSearchScope, done);
+});
+
+test('autosearch turned off does show patternfly 4 search button', async (done) => {
+  const autoSearchScope = mockSetting(nockInstance, AUTOSEARCH_WHILE_TYPING, false);
+  const autocompleteScope = mockAutocomplete(nockInstance, endpoint);
+
+  const { getByLabelText } = renderWithRedux(<Search {...props} />);
+
+  // Using waitFor as the autoSearch setting defaults to true, it won't be changed until http call
+  await waitFor(() => expect(getByLabelText(searchButtonLabel)).toBeInTheDocument());
+
+  assertNockRequest(autoSearchScope);
+  assertNockRequest(autocompleteScope, done);
+});
+
+test('autosearch turned on does not affect patternfly 3 buttons', async (done) => {
+  const autoSearchScope = mockSetting(nockInstance, AUTOSEARCH_WHILE_TYPING, true);
+  const autocompleteScope = mockAutocomplete(nockInstance, endpoint);
+
+  const { getByLabelText } = renderWithRedux(<Search {...{ ...props, patternfly4: false }} />);
+
+  await waitFor(() => expect(getByLabelText('patternfly 3 search button')).toBeInTheDocument());
+
+  assertNockRequest(autoSearchScope);
+  assertNockRequest(autocompleteScope, done);
+});
+
+test('search function is called when search is typed into with autosearch', async (done) => {
```

Comment:
Ya, I can see how this is testing more of the implementation. 

With just this component, it isn't sete picks up to update a table itself, it more does that through updating redux (in the onSearch function) or elsewhere and a tabl up the update. I added tests to the CV page tests to cover the full functionality for that page and I think we should continue to add search tests to new pages added. I wanted tests here for stability and also to make it easier to move to Foreman too.

How about I update the mocked function to return the expected object from `onSearch` and assert the API call is made? In just the context of this component, I think that falls under user expectations as that is where the UI leaves hands off to the backend API and the search is performed.

Reply:
sounds good to me! :+1:

Reply:
Looking at this again, I'm not sure how we can test anything more than the function is called. `Search` doesn't actually make the backend call to search, the action is usually dispatched in the `onSearch`  function passed in, which then calls the API, but this isn't a requirement of the component.

With `autoCompleteParams`, we can test the api call is made and component is updated, this all happens within the context of the `Search` component. But with `onSearch`, nothing is happening except calling the passed function.

It is testing an implementation detail, but its hard to avoid when just dealing with just this component. If we were to create any logic within the onSearch function and validate that, it wouldn't be testing any application code. The important thing is the function is called here and we test larger search workflows within the context of a page elsewhere, like the ones added in the CV page tests.